### PR TITLE
fix memory allocation and deallocation unmatch

### PR DESCRIPTION
**Pull request by inabao**

### CODE REVIEW

File: src/algorithm/hnswlib/hnswalg.h [hunk 1]
```diff
@@ -1329,7 +1329,7 @@ class HierarchicalNSW : public AlgorithmInterface<float> {
                 link_lists_[i] = nullptr;
             } else {
                 element_levels_[i] = link_list_size / size_links_per_element_;
-                link_lists_[i] = (char*)malloc(link_list_size);
+                link_lists_[i] = (char*)allocator_->Allocate(link_list_size);
```
Comment (jiaweizone) [inline src/algorithm/hnswlib/hnswalg.h:1332]: Do you review and replace all malloc(..) allocations?

File: src/default_allocator.cpp [hunk 2]
```diff
@@ -14,20 +14,45 @@
...
 
+#include <fmt/format.h>
+
 namespace vsag {
 void*
 DefaultAllocator::Allocate(size_t size) {
-    return malloc(size);
+    auto ptr = malloc(size);
+    std::lock_guard<std::mutex> guard(set_mutex_);
+    allocated_ptrs_.insert(ptr);
```
Comment (jiaweizone) [inline src/default_allocator.cpp:25]: Debug code ?

File: src/algorithm/hnswlib/hnswalg.h (hunk 1 above)
Comment (inabao) [reply src/algorithm/hnswlib/hnswalg.h:1332]: I had added unit tests to make sure that all the allocations and deallocations is matching.

File: src/default_allocator.cpp (hunk 2 above)
Comment (inabao) [reply src/default_allocator.cpp:25]: The code is using to ensure that the memory allocated by this allocator will be released by this allocator, while the memory not allocated by this allocator will not be released by this allocator.

File: src/default_allocator.h [hunk 3]
```diff
@@ -50,6 +59,10 @@ class DefaultAllocator : public Allocator {
 
     void*
     Reallocate(void* p, size_t size) override;
+
+private:
+    std::unordered_set<void*> allocated_ptrs_;
```
Comment (jiaweizone) [inline src/default_allocator.h:66]: add NDEBUG check

File: src/default_allocator.h [hunk 4]
```diff
@@ -33,7 +35,14 @@ class DefaultAllocator : public Allocator {
 
 public:
     DefaultAllocator() = default;
-    virtual ~DefaultAllocator() = default;
+    ~DefaultAllocator() override {
+        if (not allocated_ptrs_.empty()) {
+            logger::error(fmt::format("There is a memory leak in {}.", Name()));
+#ifndef NDEBUG
```
Comment (jiaweizone) [inline src/default_allocator.h]: move to L39 ？

**Review by jiaweizone — APPROVED** [review]
LGTM

**Pull request merged**